src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
 
   function_pointer_restrictions_json.output(outFile);
 }
+function_pointer_restrictionst function_pointer_restrictionst::merge(
+  const function_pointer_restrictionst &other) const
+{
+  return function_pointer_restrictionst{
+    merge_function_pointer_restrictions(restrictions, other.restrictions)};
+}
+
+function_pointer_restrictionst get_function_pointer_by_name_restrictions(
+  const goto_modelt &goto_model,
+  const optionst &options)
+{
+  function_pointer_restrictionst::restrictionst by_name_restrictions =
+      function_pointer_restrictionst::
+      parse_function_pointer_restrictions_from_command_line(
+        options.get_list_option(RESTRICT_FUNCTION_POINTER_BY_NAME_OPT),
+        RESTRICT_FUNCTION_POINTER_BY_NAME_OPT);
+  function_pointer_restrictionst::restrictionst restrictions;
+  for(auto const &goto_function : goto_model.goto_functions.function_map)
+  {
+    for_each_function_call(
+      goto_function.second, [&](goto_programt::const_targett location) {
+        PRECONDITION(location->is_function_call());
+        if(can_cast_expr<dereference_exprt>(
+             location->get_function_call().function()))
+        {
+          PRECONDITION(can_cast_expr<symbol_exprt>(
+            to_dereference_expr(location->get_function_call().function())
+              .pointer()));
+          auto const &function_pointer_call_site = to_symbol_expr(
+            to_dereference_expr(location->get_function_call().function())
+              .pointer());
+          auto const &body = goto_function.second.body;
+          for(auto it = std::prev(location); it != body.instructions.end();
```

Comment:
:question: Wait what? We start before the location, and then walk forward looking for an assignment? Shouldn't we be going backwards?

Reply:
Ah yes. The loop is unnecessary as the instruction we look for will always be at `std::prev(location)`. Have removed the loop.